File: benchmarks/fixtures/webapp_ts/src/events/dispatcher.ts
/** Event dispatcher. */
import { getLogger } from '../utils/helpers';

const logger = getLogger("events.dispatcher");

/** Event object. */
export interface AppEvent {
    type: string;
    data: Record<string, unknown>;
    timestamp: number;
    processed: boolean;
}

type EventHandler = (event: AppEvent) => void;

/** Central event bus. */
export class EventDispatcher {
    private handlers: Map<string, EventHandler[]> = new Map();
    private eventLog: AppEvent[] = [];

    /** Register a handler. */
    on(eventType: string, handler: EventHandler): void {
        if (!this.handlers.has(eventType)) this.handlers.set(eventType, []);
        this.handlers.get(eventType)!.push(handler);
        logger.info(`Handler registered for: ${eventType}`);
    }

    /** Emit an event. */
    emit(eventType: string, data?: Record<string, unknown>): number {
        const event: AppEvent = { type: eventType, data: data ?? {}, timestamp: Date.now(), processed: false };
        this.eventLog.push(event);
        const handlers = this.handlers.get(eventType) ?? [];
        logger.info(`Emitting ${eventType} to ${handlers.length} handlers`);
        let invoked = 0;
        for (const handler of handlers) {
            try {
                handler(event);
                invoked++;
            } catch (e) {
                logger.error(`Handler error for ${eventType}: ${e}`);
            }
        }
        event.processed = true;
        return invoked;
    }

    /** Get event count. */
    eventCount(): number {
        return this.eventLog.length;
    }
}

File: benchmarks/fixtures/webapp_rb/events/dispatcher.rb
# Event dispatcher.

require_relative '../utils/helpers'

# Central event bus.
class EventDispatcher
  def initialize
    @handlers = {}
    @event_log = []
  end

  # Register a handler for an event type.
  def on(event_type, &handler)
    @handlers[event_type] ||= []
    @handlers[event_type] << handler
    get_logger('events.dispatcher').info("Handler registered for: #{event_type}")
  end

  # Emit an event to all registered handlers.
  def emit(event_type, data = {})
    event = { type: event_type, data: data, timestamp: Time.now.to_i, processed: false }
    @event_log << event
    handlers = @handlers[event_type] || []
    get_logger('events.dispatcher').info("Emitting #{event_type} to #{handlers.length} handlers")
    invoked = 0
    handlers.each do |handler|
      begin
        handler.call(event)
        invoked += 1
      rescue StandardError => e
        get_logger('events.dispatcher').error("Handler error for #{event_type}: #{e}")
      end
    end
    event[:processed] = true
    invoked
  end

  # Get total event count.
  def event_count
    @event_log.length
  end
end
